### PR TITLE
Added raymarching reflect example.

### PR DESCRIPTION
**Pull request by gam0022**
I added an example of raymarching reflect using `ShaderMaterial`.

### CODE REVIEW

**Comment by WestLangley** [comment]
Nice!

http://ci.threejs.org/api/pullrequests/7860/examples/webgl_raymarching_reflect.html

Consider using `THREE.RawShaderMaterial`, instead, as that is all that is required.

If you do so, you will have to add
```glsl
	attribute vec3 position;
```
to the vertex shader.

File: examples/webgl_raymarching_reflect.html [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+						resolution: { type: 'v2', value: new THREE.Vector2( 512, 512 ) },
+						mouse: { type: 'v2', value: mouse },
+					},
+					vertexShader: document.getElementById( 'vertex_shader' ).textContent,
+					fragmentShader: document.getElementById( 'fragment_shader' ).textContent
+				} );
+				mesh = new THREE.Mesh( geometry, material );
+				scene.add( mesh );
+
+				renderer = new THREE.WebGLRenderer();
```
Comment (WestLangley) [inline examples/webgl_raymarching_reflect.html]: Should `renderer.setPixelRatio( window.devicePixelRatio );` be added?

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
@WestLangley Implemented your suggestions.

However... I think it would be cool if, instead of passing `mouse` and `time` to the shader we would simply pass the camera. I guess for that we have to decompose the camera matrix into position and direction?

**Comment by gam0022** [comment]
Thank you for the comment and the merging!

**Comment by WestLangley** [comment]
>I think it would be cool if, instead of passing mouse and time to the shader we would simply pass the camera. I guess for that we have to decompose the camera matrix into position and direction?

Agreed. `camera.getWorldPosition( worldPos )` and `camera.getWorldDirection( worldDir )` should do it.

@gam0022 Would you be willing to have a go at that?

**Comment by gam0022** [comment]
> Agreed. camera.getWorldPosition( worldPos ) and camera.getWorldDirection( worldDir ) should do it.
> @gam0022 Would you be willing to have a go at that?

Yes, I got it.
Do you think that I should pass cameraPos and cameraDir to the shader as type: 'v3'?

**Comment by gam0022** [comment]
Is this alright?
https://github.com/gam0022/three.js/commit/f68e1fb22daad1cfe87fbd57df55d978ffd425a7

But screen began to black out by the cameraPos...

**Comment by gam0022** [comment]
It was solved by adding dummyCamera for renderer.render().
https://github.com/gam0022/three.js/commit/f94947d200434a19c68656beb177d3360cd4b17d

Do you think it is this a good ?

File: examples/webgl_raymarching_reflect.html [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+				resolution: '512',
+			};
+
+			init();
+			render();
+
+			function init() {
+
+				scene = new THREE.Scene();
+				camera = new THREE.Camera();
```
Comment (WestLangley) [inline examples/webgl_raymarching_reflect.html]: Hmmm. I just realized you are instantiating a `THREE.Camera()` and using a simplified vertex shader. This was also done in http://threejs.org/examples/webgl_shader.html. Personally, I would use an orthographic camera to render the quad, but your approach does appear to work...

@mrdoob -- Do you have an opinion on this?

**Comment by gam0022** [comment]
I thank that there is no need to use the orthographic camera.
It's because projectionMatrix was not used in VertexShader.

**Comment by gam0022** [comment]
In  #7863, I used two cameras.

1. Camera to render the plane. (`dummyCamera`)
2. Camera in the raymarching world.(`camera`)

**Comment by WestLangley** [comment]
>I think that there is no need to use the orthographic camera.
It's because projectionMatrix was not used in VertexShader.

`gl_Position` must be in _Normalized Device Coordinate_ space. So your approach only works because, in addition to the omission of the projection matrix, the `PlaneBufferGeometry` is 2 X 2 and centered at the origin.

I assume you copied this from another example, but I expect this will be confusing to users.

@greggman Have you seen this as a common technique for rendering a full-screen quad? (simplified vertex shader, careful positioniing of the quad so the projection matrix is not required)

**Comment by greggman** [comment]
@WestLangley seems common to me. Isn't that what glslsandbox.com and shadertoy.com do? Isn't it also what http://threejs.org/examples/webgl_shader.html does? Or am I mis-understanding?

**Comment by WestLangley** [comment]
@greggman It is what `webgl_shader.html` does. I expect the OP copied that example. I am wondering if this is confusing to users -- especially since @mrdoob suggested:

>I think it would be cool if, instead of passing mouse and time to the shader we would simply pass the camera.

At this point, I am no longer sure how that suggestion makes sense in the context of this example.

**Comment by gam0022** [comment]
Thank you for the comment.

> `gl_Position` must be in Normalized Device Coordinate space. So your approach only works because, in addition to the omission of the projection matrix, the `PlaneBufferGeometry` is 2 X 2 and centered at the origin.

I understood that this approach works only in *Normalized Device Coordinate* space.

> I assume you copied this from another example, but I expect this will be confusing to users.

Yes, it is.

**Comment by gam0022** [comment]
> `gl_Position` must be in Normalized Device Coordinate space. So your approach only works because, in addition to the omission of the projection matrix, the `PlaneBufferGeometry` is 2 X 2 and centered at the origin.

Should I explain above points at comments in `webgl_raymarching_reflect.html` to avoid confusing to users?

**Comment by WestLangley** [comment]
My preference would be to use an orthographic camera. `THREE.Camera()` was not intended to be used directly -- at least that was my understanding. But since what you have implemented works, I'll leave it up to @mrdoob to decide.

I am not sure it makes sense to pass the camera "position".

**Comment by mrdoob** [comment]
> I am not sure it makes sense to pass the camera "position".

The idea was to pass a camera into the shader, so the three world is more integrated with the raymarcher world. Ideally we could pass the matrixWorld and decompose inside the shader.

**Comment by WestLangley** [comment]
Thing is, there are no objects in the scene. This is just a full-screen quad with a shader rendererd on it.

I do not feel strongly about this, but the whole thing to me is strange -- now that I think about it.

**Comment by mrdoob** [comment]
Yes, but imagine that we wanted to animate the camera. We have a lot of code for animating stuff now. We can animate the camera outside instead of replicating animation code inside the shader.

**Pull request merged**